File: java/src/test/java/io/github/linkfoundation/linksnotation/SingleLineParserTest.java
package io.github.linkfoundation.linksnotation;

import static org.junit.jupiter.api.Assertions.*;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/** Tests for single-line parsing functionality. */
class SingleLineParserTest {

  private Parser parser;

  @BeforeEach
  void setUp() {
    parser = new Parser();
  }

  @Test
  void testSingleLink() throws ParseException {
    String source = "(address: source target)";
    List<Link> links = parser.parse(source);
    String target = Link.formatLinks(links);
    assertEquals(source, target);
  }

  @Test
  void testTripletSingleLink() throws ParseException {
    String source = "(papa has car)";
    List<Link> links = parser.parse(source);
    String target = Link.formatLinks(links);
    assertEquals(source, target);
  }

  @Test
  void testBugTest1() throws ParseException {
    String source = "(ignore conan-center-index repository)";
    List<Link> links = parser.parse(source);
    String target = Link.formatLinks(links);
    assertEquals(source, target);
  }

  @Test
  void testQuotedReferences() throws ParseException {
    String source = "(a: 'b' \"c\")";
    String expected = "(a: b c)";
    List<Link> links = parser.parse(source);
    String formattedLinks = Link.formatLinks(links);
    assertEquals(expected, formattedLinks);
  }

  @Test
  void testQuotedReferencesWithSpaces() throws ParseException {
    String source = "('a a': 'b b' \"c c\")";
    // Multi-word IDs are formatted without quotes in parenthesized form (issue #184)
    String expected = "(a a: 'b b' 'c c')";
    List<Link> links = parser.parse(source);
    String formattedLinks = Link.formatLinks(links);
    assertEquals(expected, formattedLinks);
  }

  @Test
  void testParseSimpleReference() throws ParseException {
    String input = "test";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("test", result.get(0).getValues().get(0).getId());
    assertTrue(result.get(0).getValues().get(0).getValues().isEmpty());
  }

  @Test
  void testParseReferenceWithColonAndValues() throws ParseException {
    String input = "parent: child1 child2";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertEquals("parent", result.get(0).getId());
    assertEquals(2, result.get(0).getValues().size());
    assertEquals("child1", result.get(0).getValues().get(0).getId());
    assertEquals("child2", result.get(0).getValues().get(1).getId());
  }

  @Test
  void testParseMultilineLink() throws ParseException {
    String input = "(parent: child1 child2)";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertEquals("parent", result.get(0).getId());
    assertEquals(2, result.get(0).getValues().size());
  }

  @Test
  void testParseQuotedReferencesValues() throws ParseException {
    String input = "\"has space\" 'has:colon'";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(2, result.get(0).getValues().size());
    assertEquals("has space", result.get(0).getValues().get(0).getId());
    assertEquals("has:colon", result.get(0).getValues().get(1).getId());
    // Ensure formatting matches C# expectation
    assertEquals("('has space' 'has:colon')", Link.formatLinks(result));
  }

  @Test
  void testSingleLineWithId() throws ParseException {
    String input = "id: value1 value2";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testMultiLineLinkWithId() throws ParseException {
    String input = "(id: value1 value2)";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testSingletLink() throws ParseException {
    String input = "(singlet)";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("singlet", result.get(0).getValues().get(0).getId());
    assertTrue(result.get(0).getValues().get(0).getValues().isEmpty());
  }

  @Test
  void testValueLink() throws ParseException {
    String input = "(value1 value2 value3)";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testQuotedReferencesInLink() throws ParseException {
    String input = "(\"id with spaces\": \"value with spaces\")";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testSingleQuotedReferences() throws ParseException {
    String input = "('id': 'value')";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testNestedLinks() throws ParseException {
    String input = "(outer: (inner: value))";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testSpecialCharactersInQuotes() throws ParseException {
    String input = "(\"key:with:colons\": \"value(with)parens\")";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);

    input = "('key with spaces': 'value: with special chars')";
    result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testDeeplyNested() throws ParseException {
    String input = "(a: (b: (c: (d: (e: value)))))";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testHyphenatedIdentifiers() throws ParseException {
    String input = "(conan-center-index: repository info)";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testMultipleWordsInQuotes() throws ParseException {
    String input = "(\"New York\": city state)";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
    assertEquals("New York", result.get(0).getId());
  }

  @Test
  void testSimpleRef() throws ParseException {
    String input = "simple_ref";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testSimpleReferenceParser() throws ParseException {
    String input = "hello";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("hello", result.get(0).getValues().get(0).getId());
  }

  @Test
  void testQuotedReferenceParser() throws ParseException {
    String input = "\"hello world\"";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("hello world", result.get(0).getValues().get(0).getId());
  }

  @Test
  void testSingletLinkParser() throws ParseException {
    String input = "(singlet)";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("singlet", result.get(0).getValues().get(0).getId());
    assertTrue(result.get(0).getValues().get(0).getValues().isEmpty());
  }

  @Test
  void testValueLinkParser() throws ParseException {
    String input = "(a b c)";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(3, result.get(0).getValues().size());
  }

  @Test
  void testQuotedReferencesWithSpacesInLink() throws ParseException {
    String input = "(id: \"value with spaces\")";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertEquals("id", result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("value with spaces", result.get(0).getValues().get(0).getId());
  }

  @Test
  void testSingleLineWithoutId() throws ParseException {
    String input = "(value1 value2)";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(2, result.get(0).getValues().size());
  }

  @Test
  void testLinkWithId() throws ParseException {
    String input = "(id: a b c)";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertEquals("id", result.get(0).getId());
    assertEquals(3, result.get(0).getValues().size());
  }

  @Test
  void testQuotedReference() throws ParseException {
    String input = "\"quoted value\"";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(1, result.get(0).getValues().size());
    assertEquals("quoted value", result.get(0).getValues().get(0).getId());
  }

  @Test
  void testQuotedReferencesWithSpecialChars() throws ParseException {
    String input = "(\"special:char\" \"another@char\")";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertNull(result.get(0).getId());
    assertEquals(2, result.get(0).getValues().size());
    assertEquals("special:char", result.get(0).getValues().get(0).getId());
    assertEquals("another@char", result.get(0).getValues().get(1).getId());
  }

  @Test
  void testSimpleReference() throws ParseException {
    String input = "simplereference";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
  }

  @Test
  void testSingleLineLink() throws ParseException {
    String input = "id: value1 value2";
    List<Link> result = parser.parse(input);
    assertEquals(1, result.size());
    assertEquals("id", result.get(0).getId());
    assertEquals(2, result.get(0).getValues().size());
  }

  @Test
  void testSingleLineWithIdField() throws ParseException {
    String input = "myid: val1 val2";
    List<Link> result = parser.parse(input);
    assertTrue(result.size() > 0);
    assertEquals("myid", result.get(0).getId());
  }
}

File: php/tests/SingleLineParserTest.php
<?php

/**
 * Single-line parser tests - ported from the Python implementation.
 */

declare(strict_types=1);

namespace LinkFoundation\LinksNotation\Tests;

use LinkFoundation\LinksNotation\Formatter;
use LinkFoundation\LinksNotation\Parser;
use PHPUnit\Framework\TestCase;

class SingleLineParserTest extends TestCase
{
    private Parser $parser;

    protected function setUp(): void
    {
        $this->parser = new Parser();
    }

    public function testSingleLink(): void
    {
        $source = '(address: source target)';
        $this->assertSame($source, Formatter::formatLinks($this->parser->parse($source)));
    }

    public function testTripletSingleLink(): void
    {
        $source = '(papa has car)';
        $this->assertSame($source, Formatter::formatLinks($this->parser->parse($source)));
    }

    public function testBugTest1(): void
    {
        $source = '(ignore conan-center-index repository)';
        $this->assertSame($source, Formatter::formatLinks($this->parser->parse($source)));
    }

    public function testQuotedReferences(): void
    {
        $links = $this->parser->parse('(a: \'b\' "c")');
        $this->assertSame('(a: b c)', Formatter::formatLinks($links));
    }

    public function testQuotedReferencesWithSpaces(): void
    {
        $links = $this->parser->parse('(\'a a\': \'b b\' "c c")');
        $this->assertSame("('a a': 'b b' 'c c')", Formatter::formatLinks($links));
    }

    public function testParseSimpleReference(): void
    {
        $result = $this->parser->parse('test');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(1, $result[0]->values);
        $this->assertSame('test', $result[0]->values[0]->id);
        $this->assertSame([], $result[0]->values[0]->values);
    }

    public function testParseReferenceWithColonAndValues(): void
    {
        $result = $this->parser->parse('parent: child1 child2');
        $this->assertCount(1, $result);
        $this->assertSame('parent', $result[0]->id);
        $this->assertCount(2, $result[0]->values);
        $this->assertSame('child1', $result[0]->values[0]->id);
        $this->assertSame('child2', $result[0]->values[1]->id);
    }

    public function testParseMultilineLink(): void
    {
        $result = $this->parser->parse('(parent: child1 child2)');
        $this->assertCount(1, $result);
        $this->assertSame('parent', $result[0]->id);
        $this->assertCount(2, $result[0]->values);
    }

    public function testParseQuotedReferences(): void
    {
        $result = $this->parser->parse('"has space" \'has:colon\'');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(2, $result[0]->values);
        $this->assertSame('has space', $result[0]->values[0]->id);
        $this->assertSame('has:colon', $result[0]->values[1]->id);
        $this->assertSame("('has space' 'has:colon')", Formatter::formatLinks($result));
    }

    public function testParseValuesOnlyStandaloneColon(): void
    {
        // The parser is lenient here: an empty id with values is accepted
        $result = $this->parser->parse(': value1 value2');
        $this->assertNotEmpty($result);
    }

    public function testMultilineWithoutId(): void
    {
        $result = $this->parser->parse('(: value1 value2)');
        $this->assertNotEmpty($result);
    }

    public function testSingletLink(): void
    {
        $result = $this->parser->parse('(singlet)');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(1, $result[0]->values);
        $this->assertSame('singlet', $result[0]->values[0]->id);
        $this->assertSame([], $result[0]->values[0]->values);
    }

    public function testValueLink(): void
    {
        $result = $this->parser->parse('(value1 value2 value3)');
        $this->assertCount(1, $result);
        $this->assertCount(3, $result[0]->values);
    }

    public function testQuotedReferencesWithSpacesInLink(): void
    {
        $result = $this->parser->parse('("id with spaces": "value with spaces")');
        $this->assertCount(1, $result);
        $this->assertSame('id with spaces', $result[0]->id);
        $this->assertSame('value with spaces', $result[0]->values[0]->id);
    }

    public function testSingleQuotedReferences(): void
    {
        $result = $this->parser->parse("('id': 'value')");
        $this->assertCount(1, $result);
        $this->assertSame('id', $result[0]->id);
        $this->assertSame('value', $result[0]->values[0]->id);
    }

    public function testNestedLinks(): void
    {
        $result = $this->parser->parse('(outer: (inner: value))');
        $this->assertCount(1, $result);
        $this->assertSame('outer', $result[0]->id);
        $this->assertSame('inner', $result[0]->values[0]->id);
    }

    public function testSpecialCharactersInQuotes(): void
    {
        $result = $this->parser->parse('("key:with:colons": "value(with)parens")');
        $this->assertSame('key:with:colons', $result[0]->id);
        $this->assertSame('value(with)parens', $result[0]->values[0]->id);

        $result = $this->parser->parse("('key with spaces': 'value: with special chars')");
        $this->assertSame('key with spaces', $result[0]->id);
        $this->assertSame('value: with special chars', $result[0]->values[0]->id);
    }

    public function testDeeplyNested(): void
    {
        $result = $this->parser->parse('(a: (b: (c: (d: (e: value)))))');
        $this->assertCount(1, $result);
        $this->assertSame('(a: (b: (c: (d: (e: value)))))', Formatter::formatLinks($result));
    }

    public function testHyphenatedIdentifiers(): void
    {
        $result = $this->parser->parse('(conan-center-index: repository info)');
        $this->assertSame('conan-center-index', $result[0]->id);
        $this->assertCount(2, $result[0]->values);
    }

    public function testMultipleWordsInQuotes(): void
    {
        $result = $this->parser->parse('("New York": city state)');
        $this->assertNotEmpty($result);
        $this->assertSame('New York', $result[0]->id);
    }

    public function testSimpleReferenceParser(): void
    {
        $result = $this->parser->parse('hello');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(1, $result[0]->values);
        $this->assertSame('hello', $result[0]->values[0]->id);
    }

    public function testQuotedReferenceParser(): void
    {
        $result = $this->parser->parse('"hello world"');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(1, $result[0]->values);
        $this->assertSame('hello world', $result[0]->values[0]->id);
    }

    public function testValueLinkParser(): void
    {
        $result = $this->parser->parse('(a b c)');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(3, $result[0]->values);
    }

    public function testLinkWithId(): void
    {
        $result = $this->parser->parse('(id: a b c)');
        $this->assertCount(1, $result);
        $this->assertSame('id', $result[0]->id);
        $this->assertCount(3, $result[0]->values);
    }

    public function testSingleLineLink(): void
    {
        $result = $this->parser->parse('id: value1 value2');
        $this->assertCount(1, $result);
        $this->assertSame('id', $result[0]->id);
        $this->assertCount(2, $result[0]->values);
    }

    public function testQuotedReferencesWithSpecialChars(): void
    {
        $result = $this->parser->parse('("special:char" "another@char")');
        $this->assertCount(1, $result);
        $this->assertNull($result[0]->id);
        $this->assertCount(2, $result[0]->values);
        $this->assertSame('special:char', $result[0]->values[0]->id);
        $this->assertSame('another@char', $result[0]->values[1]->id);
    }
}
